### PR TITLE
Fix LSTM Layer Normalization implementation to match original paper

### PR DESCRIPTION
**Pull request by bchu**
The Layer Norm paper (https://arxiv.org/abs/1607.06450) applies layer norm to LSTMs by separately normalizing the linear transformation of the inputs and the linear transformation of the recurrent cell state. See equation (20) in the Supplementary Material (page 13).

The current contrib LSTM layer norm implementations apply layer norm to LSTMs by separately normalizing each of the gate/cell update preactivations (separate columns of the sum of the linear transformations of the inputs and recurrent cell state). This is very different from the paper (different tensors are being normalized).

This PR changes the two implementations to match the paper. There is one slight difference with the paper: I omit two of the layer norm offset _β_ parameters because there is a bias term already.

### CODE REVIEW

**Comment by caisq** [comment]
@bchu Will this introduce backward compatibility issues?

**Comment by bchu** [comment]
@caisq I believe so. It's a different implementation of layer norm in LSTMs.

**Review by ebrevdo — CHANGES_REQUESTED** [review]
Where are docstring updates?  Can you guard the different behavior behind a construction argument flag?

**Comment by bchu** [comment]
@ebrevdo I've updated the comments. I've added a flag for the different behavior (`layer_norm_columns`). Let me know if other updates are needed.

**Comment by rmlarsen** [comment]
@ebrevdo could you take another look, please?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by carlthome** [comment]
Normalizing gates individually feels nicer. Any experimental results with pros/cons of the respective methods?

Also, why not keep the offset parameters and remove the bias term instead (as in tf.contrib.layers)?

**Comment by bchu** [comment]
@carlthome I think it also makes sense from the perspective of normalizing each linear transformation separately. Presumably the recurrent states and the recurrent inputs have different distributions. I'm not aware of any experimental comparisons. I omitted the offset because that saves us an extra parameter.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: It has been 16 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: It has been 93 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: It has been 108 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: It has been 123 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by bchu** [comment]
Happy to resolve any merge conflicts if there's still interest in this PR.

**Comment by FancyCoder0** [comment]
Hi Brian Chu @bchu ,
We are researchers working on identifying redundant development and duplicated pull requests. We have found there is a pull request: https://github.com/tensorflow/tensorflow/pull/14578 which might be a potentially duplicate to this one. We would like to build the link between developers to reduce redundant development. We would really appreciate if you could help us to validate and give us feedback.
Thank you very much for your time!

**Comment by karmel** [comment]
@ebrevdo , is there still interest in this PR? @bchu and @KiewanVillatel to comment on the helpful suggestion that this might be a dupe with #14578 ?

**Comment by karmel** [comment]
Also @drpngx , who was reviewing #14578

**Comment by bchu** [comment]
#14578 implements Layer Norm for GRUs, while this PR deals with fixing the current LSTM Layer Norm implementation.

**Comment by drpngx** [comment]
Please pull rebase and push again.

I wonder if there is a way to get a cleaner terminology -- center, columns, and axis?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 20 days with no activity and the `awaiting review` label has been applied.

**Comment by ebrevdo** [comment]
Thank you for your work to fix this!  I think we should proactively take your version of the code and call it "class LayerNormLSTMCellV2"; in TensorFlow 2.0 we can get remove of the old one from the public API and call this one "LayerNormLSTMCell" in the public API.

In order to get there, I think you should cut out the pieces implement the correct version (your new code) into the new class.  Get rid of the following `__init__` arguments: scope (no longer used), layer_norm (assume true), and layer_norm_columns (assume true).  Move all variable creation to the `build` method and the computation to the `call` method.  For an example, see The [LSTMCell implementation](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell_impl.py#L781).

Do you have cycles to do this?

**Review by ebrevdo — CHANGES_REQUESTED** [review]
Please see my comments.  It would be much cleaner to just implement this in its own class as a `LayerRNNCell`.; call it `LayerNormLSTMCellV2`.

**Comment by ebrevdo** [comment]
Err (in the new cell for `layer_norm_columns` assume False: that's your code :)

**Review by AzizCode92 — APPROVED** [review]

**Comment by ebrevdo** [comment]
Is there an existing LayerNorm keras cell?  If so you can use that.
Otherwise you should abstract the layer norm code into a helper function
you can use in both.

**Comment by ebrevdo** [comment]
xref #14578 which is implementing LayerNormGRUCell and will need to have the same layer norm helper function.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by bchu** [comment]
Is this worth fixing up, or will classes like `LayerNormBasicLSTMCell` be superseded by Keras in tf 2.0?

**Comment by rthadur** [comment]
> Is this worth fixing up, or will classes like `LayerNormBasicLSTMCell` be deprecated in favor of Keras in Tensorflow 2.0?

@ebrevdo @karmel gentle ping

**Comment by qlzh727** [comment]
https://github.com/tensorflow/addons/pull/205.

The contrib is going away in TF 2.0 and we are moving selected APIs to tf addons.

I am also kind of curious about the whether the norm should be applied to the fused gates (paper version) or the individual gate separately (current implementation). I think the current implementation makes more sense since each gate carries different mathematical meaning. Unless those gates have similar numerical distribution, norm them together seems wrong to me.

**Comment by qlzh727** [comment]
https://github.com/tensorflow/addons/pull/210 has been submitted to addons.

The change includes:
1. Update the implementation to match keras style in 2.0, eg add_weights in build(), and get rid of variable_scope.
2. Implementation is now matching with the original paper. Normalization is now done with fused gates. The reference implementation from the original paper author can be found in https://github.com/ryankiros/layer-norm/blob/master/layers.py#L457-L475.
3 Fix the missing bias when normalization is enabled, which is an oversight in the original implementation.

With that, I think this PR can be closed.

Thanks for the contribution.
